### PR TITLE
Refactor/peer

### PR DESCRIPTION
**Pull request by boecklim**
- Ensure that `Shutdown()` blocks until all go routines are stopped.
- Change read retry function: Remove backoff package and ensure `wire.ReadMessage()` does not block when cancelling
- Move integration test to separate package so that docker container is only started when this test runs

### CODE REVIEW

**Review by icellan — CHANGES_REQUESTED** [review]

File: peer.go [hunk 1]
```diff
@@ -205,16 +228,22 @@ func (p *Peer) connect() error {
 		p.readConn = conn
 	}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	p.cancelWriteHandler = cancel
+	ctx := context.Background()
+
+	writerCtx, cancelWriter := context.WithCancel(ctx)
```
Comment (icellan) [inline peer.go:233]: @boecklim Shouldn't you be passing in the global context here? Even if you are creating a new one, this one should end when the global context is Done.

File: peer.go [hunk 2]
```diff
@@ -274,201 +303,242 @@ func (p *Peer) String() string {
...
-				if err != nil {
-					if errors.Is(err, context.Canceled) {
-						p.logger.Info("Retrying to read cancelled")
-						return
-					}
+	return nil, readMessageErr
+}
 
-					p.logger.Error("Retrying to read failed", slog.String(errKey, err.Error()))
+func (p *Peer) startReadHandler(cancelCtx context.Context) {
```
Comment (icellan) [inline peer.go:375]: Shouldn't this just be called `ctx` instead of `cancelCtx`? It's just context and not a cancel function.

File: peer.go [hunk 3]
```diff
@@ -589,71 +659,75 @@ func (p *Peer) writeRetry(ctx context.Context, msg wire.Message) error {
 	return backoff.RetryNotify(operation, policyContext, notify)
 }
 
-func (p *Peer) startWriteChannelHandler(ctx context.Context) {
-	p.logger.Info("Starting write handler")
+func (p *Peer) startWriteChannelHandler(cancelCtx context.Context, instance int) {
```
Comment (icellan) [inline peer.go:662]: `ctx` instead of `cancelCtx`?

File: peer.go (hunk 1 above)
Comment (boecklim) [reply peer.go]: @icellan I think, I know what you mean. Added another commit. Please take a look.

**Review by icellan — APPROVED** [review]

**Pull request merged**